File: src/main/groovy/com/moowork/gradle/node/yarn/YarnSetupTask.java
package com.moowork.gradle.node.yarn;

import com.moowork.gradle.node.NodePlugin;
import com.moowork.gradle.node.npm.NpmSetupTask;
import org.codehaus.groovy.runtime.DefaultGroovyMethods;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.OutputDirectory;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;


/**
 * Setup a specific version of Yarn to be used by the build.
 **/
public class YarnSetupTask extends NpmSetupTask {

	public YarnSetupTask() {
		this.setGroup(NodePlugin.NODE_GROUP);
		this.setDescription("Setup a specific version of Yarn to be used by the build.");

		this.setEnabled(false);
	}

	@Override
	@Input
	public Set<Object> getInput() {
		Set<Object> set = new HashSet<>();
		set.add(this.getConfig().getDownload());
		set.add(this.getConfig().getYarnVersion());
		return set;
	}

	@OutputDirectory
	public File getYarnDir() {
		return this.getVariant().getYarnDir();
	}

	@Override
	public void configureVersion(final String yarnVersion) {
		String pkg = "yarn";

		if (!yarnVersion.isEmpty()) {
			getLogger().debug("Setting yarnVersion to " + yarnVersion);
			pkg += "@" + yarnVersion;
		}

		this.setArgs(DefaultGroovyMethods.plus(DefaultGroovyMethods.plus(new ArrayList<String>(Arrays.asList("install", "--global", "--no-save")), NpmSetupTask.proxySettings()), new ArrayList<>(Arrays.asList("--prefix", this.getVariant().getYarnDir().getAbsolutePath(), pkg))));
		setEnabled(true);
	}

	public static final String NAME = "yarnSetup";
}

File: src/main/kotlin/com/moowork/gradle/node/yarn/YarnSetupTask.kt
package com.moowork.gradle.node.yarn

import com.moowork.gradle.node.NodePlugin
import com.moowork.gradle.node.npm.NpmSetupTask
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.OutputDirectory
import java.io.File
import java.util.*

/**
 * Setup a specific version of Yarn to be used by the build.
 */
open class YarnSetupTask : NpmSetupTask() {

    init {
        group = NodePlugin.NODE_GROUP
        description = "Setup a specific version of Yarn to be used by the build."
        isEnabled = false
    }

    @Input
    override fun getInput(): Set<Any?> {
        val set: MutableSet<Any?> = HashSet()
        set.add(config.download)
        set.add(config.yarnVersion)
        return set
    }

    @OutputDirectory
    fun getYarnDir(): File {
        return variant.yarnDir
    }

    override fun configureVersion(version: String) {
        var pkg = "yarn"
        if (version.isNotEmpty()) {
            logger.debug("Setting yarnVersion to $version")
            pkg += "@$version"
        }
        args = listOf("install", "--global", "--no-save", *PROXY_SETTINGS.toTypedArray(), "--prefix", variant.yarnDir.absolutePath, pkg) + args
        isEnabled = true
    }

    companion object {
        const val NAME = "yarnSetup"
    }
}
